streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java

```diff
@@ -0,0 +1,220 @@
+
+    ClientTagAwareStandbyTaskAssignor(final AssignmentConfigs configs) {
+        this.configs = configs;
+    }
+
+    @Override
+    public void assignStandbyTasks(final SortedMap<UUID, ClientState> clientStates, final Set<TaskId> statefulTasks) {
+        final int numStandbyReplicas = configs.numStandbyReplicas;
+        final Set<String> rackAwareAssignmentTags = new HashSet<>(configs.rackAwareAssignmentTags);
+        final Map<TaskId, UUID> statefulTasksWithClients = new HashMap<>();
+
+        statefulTasks.forEach(statefulTaskId -> clientStates.forEach((uuid, clientState) -> {
+            if (clientState.activeTasks().contains(statefulTaskId)) {
+                statefulTasksWithClients.put(statefulTaskId, uuid);
+            }
+        }));
+
+        final StandbyTaskDistributor standbyTaskDistributor = new StandbyTaskDistributor(
+            numStandbyReplicas,
+            clientStates,
+            rackAwareAssignmentTags,
+            statefulTasksWithClients
+        );
+
+        statefulTasksWithClients.forEach(standbyTaskDistributor::assignStandbyTasksForActiveTask);
+    }
+
+    @Override
+    public boolean isValidTaskMovement(final TaskMovementAttempt taskMovementAttempt) {
+        final Map<String, String> sourceClientTags = taskMovementAttempt.sourceClient().clientTags();
+        final Map<String, String> destinationClientTags = taskMovementAttempt.destinationClient().clientTags();
+
+        for (final Entry<String, String> sourceClientTagEntry : sourceClientTags.entrySet()) {
+            if (!sourceClientTagEntry.getValue().equals(destinationClientTags.get(sourceClientTagEntry.getKey()))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static final class StandbyTaskDistributor {
+        private final int numStandbyReplicas;
+        private final SortedMap<UUID, ClientState> clientStates;
+        private final Set<String> rackAwareAssignmentTags;
+        private final Map<String, Set<String>> tagKeyToTagValuesMapping;
+        private final Map<String, Set<UUID>> clientsPerTagValue;
+        private final ConstrainedPrioritySet standbyTaskClientsByTaskLoad;
+
+        private final Set<UUID> usedClients = new HashSet<>();
+        private final Map<TaskId, Integer> tasksToRemainingStandbys;
+
+        StandbyTaskDistributor(final int numStandbyReplicas,
+                               final SortedMap<UUID, ClientState> clientStates,
+                               final Set<String> rackAwareAssignmentTags,
+                               final Map<TaskId, UUID> statefulTasksWithClients) {
+            tagKeyToTagValuesMapping = new HashMap<>();
+            clientsPerTagValue = new HashMap<>();
+            standbyTaskClientsByTaskLoad = new ConstrainedPrioritySet(
+                (client, t) -> !clientStates.get(client).hasAssignedTask(t),
+                client -> clientStates.get(client).assignedTaskLoad()
+            );
```

Comment:
Would it be possible to integrate the tag constraint as part of the constraint on the priority queue?

Reply:
It's already part of the poll constraint of the priority queue. Example:

```
final UUID polledClient = standbyTaskClientsByTaskLoad.poll(
    activeTaskId, uuid -> !clientsOnAlreadyUsedTagDimensions.contains(uuid)
);
```
I don't think it will be doable with constructor constraint, because we need to update constraint on each poll.